### PR TITLE
Jv thunder convnext final

### PR DESCRIPTION
**Pull request by JvThunder**

### CODE REVIEW

**Review by sezan92 — DISMISSED** [review]
I think it is better to add a separate directory for images. like

```
chapters/en/Unit 2 - Convolutional Neural Networks/images/convnext/block_comparison.png
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 1]
```diff
@@ -0,0 +1,82 @@
+
+# ConvNext - A ConvNet for the 2020s (2022)
+
+## Introduction
+The paper "A ConvNet for the 2020s" proposed the ConvNext architecture in 2022 by a team of researchers from Facebook AI Research, including Zhuang Liu, Hanzi Mao, Chao-Yuan Wu, Christoph Feichtenhofer, Trevor Darrell, and Saining Xie.
```
Comment (sezan92) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]: the paper details can be added to the Reference section

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+Recently, the breakthrough of Vision Transformers (ViTs) quickly superseded pure ConvNet models as the new state-of-the-art for image recognition.
```
Comment (sezan92) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
Recently, the breakthrough of Vision Transformers (ViTs) quickly superseded pure CNN models as the new state-of-the-art for image recognition.
```

**Review by sezan92 — COMMENTED** [review]
commented. I will add more

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx (hunk 1 above)
Comment (JvThunder) [reply chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]: changed

**Review by themurtazanazir — APPROVED** [review]
looks good. as a suggestion, add a bit more details or explanations. It seems a bit too fast moving to me at points. But other than that, it looks great. I enjoyed the read. Thanks

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+ConvNext starts with a regular ResNet (ResNet-50) and gradually modernizes and modifies the architecture to closely imitate the hierarchical structure of Vision Transformers.
+The key improvements are:
+- Training Techniques
+- Macro Design
+- ResNeXt-ify
+- Inverted Bottleneck
+- Large Kernel Sizes
+- Micro Design
+We will go through each of the key improvements briefly.
+These designs are not novel in itself, however you can learn how researchers adapt and modify designs systematically to improve existing models.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
These designs are not novel in itself, however, you can learn how researchers adapt and modify designs systematically to improve existing models.
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+ConvNext applies training techniques close to DeiT and Swin Transformers which are:
+- Extended epochs from the original 90 epochs to 300 epochs
+- AdamW optimizer
+- Data augmentation techniques such as Mixup, Cutmix, RandAugment, Random Erasing
+- Regularization schemes including Stochastic Depth and Label Smoothing
+This improves ResNet-50's accuracy from 76.1% to 78.8%.
+
+
+## Macro Design
+Analyzing Swin Transformers' macro network, there are two interesting design considerations that is used in ConvNext.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
Analyzing Swin Transformers' macro network, two interesting design considerations are used in ConvNext.
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+### The stage compute ratio
+The stage compute ratio refers to how the computational load is distributed among the stages of a neural network model.
+Following Swin Transformer's compute ratio of 1:1:3:1, the researchers adjusted the number of blocks on each stage of the ResNet from the original (3, 4, 6, 3) to (3, 3, 9, 3). 
+This improves the model accuracy from 78.8% to 79.4%.
+
+### Changing stem to Patchify
+Typically, on the start of ResNet's architecture, the input is fed to a stem a 7×7 convolution layer with stride 2, followed by a max pool, which is used to downsample the image by a factor of 4.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
Typically, at the start of ResNet's architecture, the input is fed to a stem a 7×7 convolution layer with stride 2, followed by a max pool, which is used to downsample the image by a factor of 4.
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+However, it turns out to be substitutable to a simpler but more effective 4×4 non-overlapping convolutional layer with stride 4.
+This improves the model accuracy from 79.4% to 79.5%.
+
+
+## ResNeXt-ify
+ConvNext adopts the idea of ResNeXt, which has been explored in the previous page. 
+In short, ResNeXt has a better FLOPs (number of floating point operations) over accuracy trade-off than a vanilla ResNet by having grouped convolution for the 3×3 conv layer in the bottleneck block.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
In short, ResNeXt has better FLOPs (number of floating point operations) over accuracy trade-off than a vanilla ResNet by having grouped convolution for the 3×3 conv layer in the bottleneck block.
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+Specifically, ConvNext increases the network width to 96 channels.
+This improves the model accuracy from 79.5% to 80.5%.
+
+
+## Inverted Bottleneck
+An important idea in every Transformer block is usage of inverted bottleneck, where the hidden layers are bigger (by a ratio of 4 for ConvNext) than the input dimension.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
An important idea in every Transformer block is the usage of an inverted bottleneck, where the hidden layers are bigger (by a ratio of 4 for ConvNext) than the input dimension. 
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+## Inverted Bottleneck
+An important idea in every Transformer block is usage of inverted bottleneck, where the hidden layers are bigger (by a ratio of 4 for ConvNext) than the input dimension. 
+This idea has also been used and popularized in Computer Vision by MobileNetV2.
+By using this technique, it improves the model accuracy from 80.5% to 80.6%.
+[Inverted Bottleneck Comparison](./images/convnext/inverted_bottleneck.png)
+
+
+## Large Kernel Sizes
+One of the reasons Vision Transformer do so well has to do with its non-local self-attention which enables a more global receptive field of image features.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
One of the reasons Vision Transformer does so well is its non-local self-attention which enables a more global receptive field of image features.
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+- With the prerequisite set up, we can get the benefit of adopting larger kernel-sized convolutions. 
+By using kernel size 7x7, the network retains the accuracy while improving the model's overall FLOPs.
+[Moving up the Depth Conv Layer](./images/convnext/depthwise_moveup.png)
+
+## Micro Design
+In addition to the modifications stated, ConvNext also considers micro design, focusing on layer level details.
+The changes are:
+- Replacing ReLU activation with GELU (Gaussian Error Linear Unit) (optional, accuracy stays unchanged) 
+- Fewer activation functions (eliminate all GELU layers from the residual block except for one between two 1×1 layers, replicating the style of a Transformer block)
+- Fewer normalization layers (remove two BatchNorm layers, leaving only one BN layerbefore the conv 1 × 1 layers)
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
- Fewer normalization layers (remove two BatchNorm layers, leaving only one BN layer before the conv 1 × 1 layers)
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+- Separate downsampling layers (add a separate downsample layer in between ResNet stages, adding LN layers before each downsampling layer)
+With these final modifications, this improves the final model accuracy from 80.6% to 82.0%, exceeding Swin Transformer's accuracy (81.3%).
+
+
+## Model Code
+You can go to [this HuggingFace documentation](https://huggingface.co/docs/transformers/model_doc/convnext) to learn more about how to integrate the ConvNext pipeline into your code.
+
+## References
+The paper "A ConvNet for the 2020s" proposed the ConvNext architecture in 2022 by a team of researchers from Facebook AI Research, including Zhuang Liu, Hanzi Mao, Chao-Yuan Wu, Christoph Feichtenhofer, Trevor Darrell, and Saining Xie.
+The paper can be found [here](https://arxiv.org/abs/2201.03545) and the github repo can be found [here](https://github.com/facebookresearch/ConvNeXt).
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
The paper can be found [here](https://arxiv.org/abs/2201.03545) and the GitHub repo can be found [here](https://github.com/facebookresearch/ConvNeXt).
```

**Review by sitammeur — DISMISSED** [review]
A well-written, information-rich one. I made a few little recommendations here.

**Review by sezan92 — APPROVED** [review]

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+To show the effectiveness of each improvement, we will compare the model's accuracy before and after the modification on ImageNet-1K.
+[Block Comparison](./images/convnext/block_comparison.png)
+
+
+## Training Techniques
+The researchers first discerned that, while architectural design choices are crucial, the quality of the training procedure also plays a pivotal role in influencing performance outcomes.
+Inspired by DeiT and Swin Transformers, ConvNext closely adapts their training techniques. Some of the notable changes are:
+- Epochs: Extending the epochs from the original 90 epochs to 300 epochs.
+- Optimizer: Using AdamW optimizer instead of Adam optimizer, which differs in how it handles weight decay.
+- Data augmentation: Using Mixup, Cutmix, RandAugment, Random Erasing to increase training data.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
- Data augmentation: Using Mixup, Cutmix, RandAugment, and Random Erasing to increase training data.
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+- Regularization: Using Stochastic Depth and Label Smoothing as regularization techniques.
+Modifying these training procedures has improved ResNet-50's accuracy from 76.1% to 78.8%.
+
+
+## Macro Design
+Macro design refers to the high-level structural decisions and considerations made in a system or model, such as the arrangement of layers, the distribution of computational load across different stages, and the overall structure.
+Examining Swin Transformers' macro network, the authors have identified two noteworthy design considerations that benefits the performance of ConvNext.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
Examining Swin Transformers' macro network, the authors have identified two noteworthy design considerations that benefit the performance of ConvNext.
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+### The stage compute ratio
+The stage compute ratio refers to the distribution of computational load among the stages of a neural network model.
+ResNet-50 has four main stages with (3, 4, 6, 3) blocks, which means it has a compute ratio of 3:4:6:3.
+To follow Swin Transformer's compute ratio of 1:1:3:1, the researchers adjusted the number of blocks on each stage of the ResNet from (3, 4, 6, 3) to (3, 3, 9, 3) instead. 
+Changing the stage compute ratio improves the model accuracy from 78.8% to 79.4%.
+
+### Changing stem to Patchify
+Typically, at the start of ResNet's architecture, the input is fed to a stem 7×7 convolution layer with stride 2, followed by a max pool, used to downsample the image by a factor of 4.
+However, the authors discovered that substituting the stem with a convolutional layer featuring a 4×4 kernel size and a stride of 4 is more effective, effectively convolving them through non-overlapping 4x4 patches. 
+Patchify serves the same purpose of downsampling the image by a factor of 4, while reducing the number of layers.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
Patchify serves the same purpose of downsampling the image by a factor of 4 while reducing the number of layers.
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+This Patchifying step slightly improves the model accuracy from 79.4% to 79.5%.
+
+
+## ResNeXt-ify
+ConvNext also adopts the idea of ResNeXt, explained in the previous section. 
+ResNeXt demonstrates an improved trade-off between the number of floating-point operations (FLOPs) and accuracy compared to a standard ResNet.
+By using depthwise convolution and 1 × 1 convolutions, we would have the separation of spatial and channel mixing - a characteristic also found in vision Transformers.
+Using depthwise convolutions reduces the number of FLOPS and the accuracy.
+However, by increase the channels from 64 to 96, the accuracy is higher than the original ResNet-50 while maintaining a similar number of FLOPs.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
However, by increasing the channels from 64 to 96, the accuracy is higher than the original ResNet-50 while maintaining a similar number of FLOPs.
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+By using this technique, it improves the model accuracy from 80.5% to 80.6%.
+[Inverted Bottleneck Comparison](./images/convnext/inverted_bottleneck.png)
+
+
+## Large Kernel Sizes
+A key factor contributing to the exceptional performance of Vision Transformer is its non-local self-attention, allowing for a broader receptive field of image features.
+In Swin Transformers, the attention block window size is set to at least 7×7, surpassing the 3x3 kernel size of ResNext. 
+However, before adjusting the kernel size, it is necessary to reposition the depthwise convolution layer, as shown in the image below. 
+This repositioning enables the 1x1 layers to efficiently handle computational tasks, while the depthwise convolution layer functions as a more non-local receptor.
+With this, the network can harness the advantages of incorporating a bigger kernel-sized convolutions.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
With this, the network can harness the advantages of incorporating bigger kernel-sized convolutions. 
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+In Swin Transformers, the attention block window size is set to at least 7×7, surpassing the 3x3 kernel size of ResNext. 
+However, before adjusting the kernel size, it is necessary to reposition the depthwise convolution layer, as shown in the image below. 
+This repositioning enables the 1x1 layers to efficiently handle computational tasks, while the depthwise convolution layer functions as a more non-local receptor.
+With this, the network can harness the advantages of incorporating a bigger kernel-sized convolutions. 
+Implementing a 7x7 kernel size maintains the accuracy at 80.6% but reduces the overall FLOPs efficiency of the model.
+[Moving up the Depth Conv Layer](./images/convnext/depthwise_moveup.png)
+
+
+## Micro Design
+In addition to the modifications stated, the author adds some micro design changes to the model.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
In addition to the modifications stated, the author adds some micro-design changes to the model.
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+Mico design refers to the low-level structural decisions such as the choice of activation functions and layers details.
```
Comment (sitammeur) [inline chapters/en/Unit 2 - Convolutional Neural Networks/convnext.mdx]:
```suggestion
Mico design refers to low-level structural decisions such as the choice of activation functions and layer details.
```

**Review by sitammeur — CHANGES_REQUESTED** [review]
For pictures, share them on Discord. I will upload it to the hugging face course asset.

**Review by themurtazanazir — APPROVED** [review]

**Review by albertkao227 — APPROVED** [review]

**Pull request merged**